src-java/stats-topology/stats-storm-topology/src/main/java/org/openkilda/wfm/topology/stats/service/KildaEntryCacheService.java

```diff
@@ -302,4 +414,138 @@ private static KildaEntryDescriptor newEndpointPathEntry(
+        int yPointSwitchPosition = findYPointSwitchPosition(switchIds, yPointSwitchId);
+
+        for (int i = 0; i < path.size(); i++) {
+            boolean isShared = isSharedSwitch(i, yPointSwitchPosition, cookie);
+            SwitchId sw = path.get(i).getSwitchId();
+
+            FlowSegmentCookie modifiedCookie = isShared
+                    ? cookie.toBuilder().subType(FlowSegmentCookie.FlowSubType.SHARED).build() : cookie;
+            //ingress
+            if (i == 0) {
+                cacheHandler.handle(newHaFlowDescriptor(sw, INGRESS, haFlowId, modifiedCookie,
+                        meterId, null, null, resolveHaSubFlowId(isShared, haSubFlowId)));
+            }
+            //y-flow
+            if (sw.equals(yPointSwitchId)) {
+                cacheHandler.handle(newHaFlowDescriptor(sw, HA_FLOW_Y_POINT, haFlowId, modifiedCookie, null,
+                        ypointGroupId, yPointMeterId, resolveHaSubFlowIdForYPoint(cookie, haSubFlowId)));
+            } else if (i > 0 && i < path.size() - 1) {
+                // do not send meter into a transit
+                cacheHandler.handle(newCommonHaFlowDescriptor(sw, TRANSIT, haFlowId, modifiedCookie,
+                        null, null, resolveHaSubFlowId(isShared, haSubFlowId)));
+            }
+            //egress
+            if (i == path.size() - 1) {
+                cacheHandler.handle(newHaFlowDescriptor(sw, EGRESS, haFlowId, modifiedCookie,
+                        null, null, null, resolveHaSubFlowId(isShared, haSubFlowId)));
+            }
+        }
+    }
+
+    private String resolveHaSubFlowId(boolean isShared, String haSubFlowId) {
+        return isShared ? HA_SUBFLOW_ID_SHARED : haSubFlowId;
+    }
+
+    private String resolveHaSubFlowIdForYPoint(FlowSegmentCookie cookie, String haSubFlowId) {
+        return cookie.getDirection() == FlowPathDirection.FORWARD ? HA_SUBFLOW_ID_SHARED : haSubFlowId;
+    }
+
+    /**
+     * Checks if a switch is a shared switch based on its position in the path and the yPointSwitchPosition.
+     * If switch is a shared switch then it means that it belongs to both subFlows
+     *
+     * @param switchPositionInPath the position of the switch in the path
+     * @param yPointSwitchPosition the position of the yPointSwitch in the path
+     * @param cookie               the FlowSegmentCookie associated with the switch
+     * @return true if the switch is a shared switch, false otherwise
+     */
+    private boolean isSharedSwitch(int switchPositionInPath, int yPointSwitchPosition, FlowSegmentCookie cookie) {
+        if (switchPositionInPath < 0 || cookie.getDirection() == FlowPathDirection.UNDEFINED) {
+            return false;
+        }
+        if (cookie.getDirection() == FlowPathDirection.FORWARD) {
+            return switchPositionInPath <= yPointSwitchPosition;
+        } else {
+            return switchPositionInPath > yPointSwitchPosition;
```

Comment:
the possibility of yPointSwitchPosition to be -1 seems strange to me. It makes a valid case when all switches are shared, but there is no y-point. Is that intended for isSharedSwitch to return true when findYPointSwitchPosition returns -1? Isn't it that no y-point means illegal state?